sympy/polys/matrices/normalforms.py

```diff
@@ -30,12 +37,35 @@ def smith_normal_form(m):
     ...                   [ZZ(2), ZZ(16), ZZ(14)]], (3, 3), ZZ)
     >>> print(smith_normal_form(m).to_Matrix())
     Matrix([[1, 0, 0], [0, 10, 0], [0, 0, -30]])
-
     '''
-    invs = invariant_factors(m)
-    smf = DomainMatrix.diag(invs, m.domain, m.shape)
-    return smf
-
+    '''
+    Return the Smith Normal Form of a matrix `m` along with the unimodular matrices S and T.
+    '''
+    from sympy import Matrix
+
+    M = Matrix(m)
+    S = Matrix.eye(M.rows)
+    T = Matrix.eye(M.cols)
+    J = M.copy()
+    for i in range(min(J.rows, J.cols)):
+        while J[i,i] != J[:i+1,i+1:].applyfunc(lambda x: x % J[i,i]).det():
+            non_zero_elements = [x for x in J[:,i].tolist() if x[0]]
+            g, s, t = extended_gcd(J[i,i], min(non_zero_elements))
```

Comment:
Does this function really work with all principal ideal domain, than integers?

Reply:
There are examples of polynomial matrix in 

https://en.wikipedia.org/wiki/Smith_normal_form#Similarity